File: Convert Sorted Array to Binary Search Tree/Convert Sorted Array to Binary Search Tree.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode sortedArrayToBST(int[] num) {
        int len = num.length;
        if(len == 0){
            return null;
        }
        int val = num[len / 2];
        TreeNode newNode = new TreeNode(val);
        
        if(len / 2 <= 0){
            newNode.left = null;
        }
        else{
            int[] newNum = Arrays.copyOfRange(num, 0, len / 2);
            newNode.left = sortedArrayToBST(newNum);
        }
        
        if(len / 2 + 1 >= len){
            newNode.right = null;
        }
        else{
            int[] newNum = Arrays.copyOfRange(num, len / 2 + 1, len);
            newNode.right = sortedArrayToBST(newNum);
        }
        
        return newNode;
    }
}

File: Convert Sorted Array to Binary Search Tree/Convert Sorted Array to Binary Search Tree.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:

    TreeNode* recur(vector<int>& num, int left, int right){

        int mid = (left + right) / 2;
        TreeNode* newNode = new TreeNode(num[mid]);

        if(left == right){
            newNode->left = NULL;
            newNode->right = NULL;
        }
        else if(left == mid){
            newNode->left = NULL;
            newNode->right = recur(num, mid + 1, right);
        }
        else if(right == mid){
            newNode->left = recur(num, left, mid - 1);
            newNode->right = NULL;
        }
        else{
            newNode->left = recur(num, left, mid - 1);
            newNode->right = recur(num, mid + 1, right);
        }
        
        return newNode;
    }

    TreeNode *sortedArrayToBST(vector<int> &num) {
        if(num.empty())    return NULL;
        TreeNode* res = recur(num, 0, num.size() - 1);
        return res;
    }
};
